### ISSUE TITLE
How to write Owned(T) ownership transfer arguments?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#8939]
Right now, we write ownership transfer of an `Owned` with the `in` intent:

``` chapel
proc takesOwnedThing( in arg:Owned(MyClass)) {
}
var myOwned = new Owned(new MyClass());
takesOwnedThing(myOwned);
writeln(myOwned); // outputs nil, or compiler error - myOwned is empty
```

Is this sufficient? What about the case when `takesOwnedThing` is generic?
``` chapel
proc takesOwnedThing(in arg) {
}
var myOwned = new Owned(new MyClass());
takesOwnedThing(myOwned);
writeln(myOwned); // outputs nil, or compiler error - myOwned is empty
```

Now it might be surprising that `in arg` causes the ownership transfer, since it doesn't say anything about `owned`. Additionally, it might be surprising since a borrow will have `const in` intent by default, and perhaps the default intent for Owned is some sort of borrow (see #8618).

We might consider requiring that the type modifier `owned` be included explicitly in the argument declaration to enable ownership transfer:

``` chapel
proc takesOwnedThing(in arg:owned) {
}
```

Or we might opt to use a totally different (new) argument intent, e.g.
``` chapel
proc takesOwnedThing(move arg) {
}
```

Or, we might leave `in` out of the ownership transfer business and use the `ref` intent and then transfer ownership inside of the function, e.g.
``` chapel
proc takesOwnedThing(ref arg) {
  var takeArgument = arg; // copy-init transfers ownership
}
```

or
``` chapel
proc takesOwnedThing(ref arg) {
  var takeArgument = move(arg);
}
```

**Comment by mppf** [comment]
My Hypothesis:  it would be better to require generic code doing ownership transfer to explicitly say 'owned' along with the argument. This prevents certain generic code from being reused for integers as well as owned instances, but such generic code wouldn't work for owned things without thought on the part of the developer

**Comment by mppf** [comment]
It's interesting to note that for `owned`, it might be appealing to have a `move` argument intent. However, one would not use a `move` intent for `shared` and there I think the `in` intent is more reasonable. And it seems odd to me that writing an argument that extends/transfers ownership would have a very different form in these two cases.

**Comment by vasslitvinov** [comment]
Transferring ownership upon `in` intent is a natural parallel to assignment. Although it took me a while to "get" that.

If we introduce a `move` intent, I would argue for requiring `move` instead of an assignment as well.

If so, I would allow `move` on other types as well, like records and integers, with the semantics of normal assignment. For "shared" things, I guess we would allow assignment meaning "share it" and `move` meaning "take over from the r.h.s.", perhaps with "while preserving all other owners if any".

**Comment by psahabu** [comment]
Responding to the different ideas in order:
- `in arg` works for me, but its interaction with the default intent of borrow might be confusing.
- `in arg: owned` helps with the confusion around `in arg`.
- `move` intent puts me off a bit because it introduces new syntax.
- Both `ref arg` proposals were initially appealing, but functions that do an ownership transfer wouldn't be self-documenting, hampering users of a library/API.

Taken together, I prefer `in arg`, going with `in arg: owned` if others prefer it for clarity.

**Comment by mppf** [comment]
> If we introduce a move intent, I would argue for requiring move instead of an assignment as well.

What would that look like?
``` chapel
var x move= y;
var x = move(y);
```
? Arguably the second case is "assignment" from a syntactical point of view but I don't know if you'd rule that out.

> move intent puts me off a bit because it introduces new syntax.

I don't think that's sufficient grounds to not do it. The owned/borrowed/raw changes are going to be big and noisy already. If it's adding new *useless* syntax, that's a different matter.

> Both ref arg proposals were initially appealing, but functions that do an ownership transfer wouldn't be self-documenting, potentially hampering users of a library/API.

And yet a function taking an `Owned` in by `ref` could always do ownership transfer. I suppose you are saying that you miss the way that the intent can convey how the argument behaves, in the case of `Owned`. I'm not sure that's "worth it" given the potential confusion with `in` (especially if the default intent for Owned is to "borrow the class pointer by in intent"). However, I don't yet know how we'd write that a `shared` should increase its reference count (instead of borrow), when passed by ref, without breaking the symmetry between `in` intent and variable initialization.

E.g. if we were to make the `in` intent still borrow, we might use the `ref` intent like this:
``` chapel
proc keepsSharedThing(ref arg) {
  var keptArgument = arg.share();
}
proc takesOwnedThing(ref arg) {
  var takeArgument = move(arg);
}
```

**Comment by mppf** [comment]
Closing this because it's implemented (per #8938)
